syntax = "proto3";

package mlservice;

// Основной сервис gRPC
service MLService {
  // Статус сервиса
  rpc HealthCheck (HealthRequest) returns (HealthResponse);

  // Список доступных классов моделей
  rpc ListModelClasses (Empty) returns (ModelClassesResponse);

  // Список датасетов
  rpc ListDatasets (Empty) returns (DatasetListResponse);

  // Список моделей
  rpc ListModels (Empty) returns (ModelListResponse);

  // Обучение новой модели
  rpc TrainModel (TrainModelRequest) returns (TrainModelResponse);

  // Переобучение существующей модели
  rpc RetrainModel (RetrainModelRequest) returns (TrainModelResponse);

  // Удаление модели
  rpc DeleteModel (DeleteModelRequest) returns (DeleteModelResponse);

  // Инференс
  rpc Predict (PredictRequest) returns (PredictResponse);
}

// Пустое сообщение
message Empty {}

// ---------------- Системные сообщения ----------------

message HealthRequest {}

message HealthResponse {
  string status = 1;
  string version = 2;
}

// ---------------- Классы моделей ----------------

message ModelClassInfo {
  string id = 1;
  string description = 2;
  // default_hyperparams сериализуем как JSON-строку
  string default_hyperparams_json = 3;
}

message ModelClassesResponse {
  repeated ModelClassInfo classes = 1;
}

// ---------------- Датасеты ----------------

message DatasetInfo {
  int32 id = 1;
  string name = 2;
  string path = 3;
  string description = 4;
  string version = 5;
}

message DatasetListResponse {
  repeated DatasetInfo datasets = 1;
}

// ---------------- Модели ----------------

message TrainModelRequest {
  string name = 1;
  string model_class = 2;
  int32 dataset_id = 3;
  // Гиперпараметры — JSON-строка
  string hyperparams_json = 4;
}

message RetrainModelRequest {
  int32 model_id = 1;
  string model_class = 2;
  string hyperparams_json = 3;
}

message TrainModelResponse {
  int32 model_id = 1;
  string status = 2;
}

message ModelInfo {
  int32 id = 1;
  string name = 2;
  string model_class = 3;
  int32 dataset_id = 4;
  string status = 5;
  string hyperparams_json = 6;
  string clearml_model_id = 7;
}

message ModelListResponse {
  repeated ModelInfo models = 1;
}

message DeleteModelRequest {
  int32 model_id = 1;
}

message DeleteModelResponse {
  string status = 1;
}

// ---------------- Инференс ----------------

message FloatVector {
  repeated double values = 1;
}

message PredictRequest {
  int32 model_id = 1;
  repeated FloatVector features = 2;
}

message PredictResponse {
  repeated int32 predictions = 1;
}
